swarm/src/lib.rs

```diff
@@ -603,38 +614,54 @@ where
                 Poll::Pending => network_not_ready = true,
                 Poll::Ready(NetworkEvent::ConnectionEvent { connection, event }) => {
                     let peer = connection.peer_id();
-                    let connection = connection.id();
-                    this.behaviour.inject_event(peer, connection, event);
+                    let conn_id = connection.id();
+                    if this.banned_peer_connections.contains(&conn_id)
+                        || this.banned_peers.contains(&peer)
```

Comment:
Why do we do the `|| this.banned_peers.contains(&peer)` check? 

My guess: So that if a peer banned we don't `inject_event` even on connections from before the ban. This is done because the disconnect of existing connections takes a bit and we don't want to get events from banned peers.

Reply:
That's the reason. That, and every possible condition of peers being banned, unbanned, etc and connections not closing on time

Reply:
At the end, it's also the safer bet

Reply:
Sorry in advance for reviving this discussion once more, especially since you already tested this version on your infra @divagant-martian.

Imagine the scenario below:

- Local peer is connected to peer A.
- Peer A sends two events to the local peer. The second event depends on the first event. The events are about to be delivered to the local `Swarm` from the local `Network`.
- User banns peer A.
- User polls the `Swarm`.
  - The `Swarm` receives the first event from `Network`.
  - `Swarm` drops it due to `this.banned_peers.contains(&peer)`.
- User unbanns A.
- User polls the `Swarm`.
  - The `Swarm` receives the second event from `Network`.
  - The `Swarm` does not drop the second event as `this.banned_peers.contains(&peer)` is no longer true.
  - `Swarm` delivers the second event to the `NetworkBehaviour`.

Result: `NetworkBehaviour` receives the second but not the first event from peer A.

I don't think the behavior above is intuitive for an implementor of `NetworkBehaviour`. In other words, as an implementor of `NetworkBehaviour` I would expect to either receive all events from a single remote peer in their correct order or not at all.

With the above in mind, would it not be better to remove `|| this.banned_peers.contains(&peer)` @divagant-martian and @MarcoPolo? A `NetworkBehaviour` would potentially receive events from banned peers until already existing connections are closed, but a `NetworkBehaviour` would never receive events _out of order_.

Reply:
addressed in b43a169